z3c/dependencychecker/db.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class ImportsDatabase(object):
+
+    def __init__(self):
+        self._requirements = []
+        self._extras_requirements = {}
+
+    def set_requirements(self, requirements_list):
+        for requirement in requirements_list:
+            if self._requirement_already_exists(requirement):
+                logger.warn(
+                    '"%s" is declared twice on setup.py',
+                    requirement.name,
+                )
+            else:
+                self._requirements.append(requirement)
+
+    def _requirement_already_exists(self, requirement):
+        names = self.get_requirements_names()
+        return bool(requirement.name in names)
+
+    def get_requirements_names(self):
+        return [
+            x.name
+            for x in self._requirements
+        ]
+
+    def set_extra_requirements(self, extra_name, dotted_names):
+        """A bit of extra work needs to be done as pkg_resources API returns
+        all common requirements as well as the extras when asking for an extra
+        requirements.
+        """
+        only_extra_dotted_names = self._remove_duplicates(dotted_names)
+        if extra_name in self.get_extras():
+            new_dotted_names = self._only_new_requirements(
+                extra_name,
+                dotted_names,
+            )
+            self._extras_requirements[extra_name] += new_dotted_names
+
+            logger.warn(
+                'extra requirement "%s" is declared twice on setup.py',
+                extra_name,
+            )
+        else:
+            self._extras_requirements[extra_name] = only_extra_dotted_names
+
+    def _remove_duplicates(self, imports, other=None):
+        other_names = other
+        if other is None:
+            other_names = self.get_requirements_names()
+
+        filtered_list = [
+            x
+            for x in imports
+            if x.name not in other_names
+        ]
+
+        return filtered_list
+
+    def _only_new_requirements(self, extra_name, dotted_names):
+        current_names = self.get_requirements_names_for_extra(extra_name)
+        only_new = self._remove_duplicates(
+            dotted_names,
+            other=current_names,
+        )
+        return only_new
+
+    def get_extras(self):
```

Comment:
In this case, the "get" name really confuses me a bit when reading the code. It sounds like you're downloading some extra stuff... It is probably the combination with "extra" that is the problem.

Do you need this one-liner function? Perhaps you can just say `self._extras_requirements.keys()` in the one or two places where you need it?

Reply:
That's probably due to me reading the excellent [refactoring book](https://martinfowler.com/books/refactoring.html) where the author puts a lot of emphasis in isolating the code and not leaking implementation details where they do not belong (he goes as far as to say that methods should have at maximum 3 statements, no more, and if you can, less if possible.

He's reasoning is that all statements in a method should have the same level of specificity, i.e.

```
self.get_numbers()
self.calculate_median()
self.print_report()
```

In his opinion, even if calculate_median is just one or two statements long, it should still be in a separate method rather than intermixed with highlevel code.